HDDS-11389. Incorrect number of deleted containers shown in Recon UI.

ArafatKhan2198 opened the pull request:
## What changes were proposed in this pull request?
The problem was that Recon wasn't accurately updating the state of containers, especially when they were deleted in SCM or had an `EMPTY_MISSING` unhealthy state. The fix involved adding logic to ensure that when SCM marks a container as `DELETED` or identifies it as `EMPTY_MISSING`, Recon correctly updates its state, with detailed logging to track these changes and clear documentation to help future developers understand the process.
## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11389
## How was this patch tested?

- Tested it out manually on a docker cluster 
- Made sure the existing test cases passed.
- Also added my own Unit test to make sure the updateContainerState() is called for containers deleted in SCM.

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -273,34 +278,44 @@ private long processExistingDBRecords(long currentTime,
+          }
+
+          // If the container is in the EMPTY_MISSING state, delete the record
+          if (currentContainer.isEmptyMissing()) {
```
Do empty missing needs verified from SCM ? missing is possible if DN is not registered and same can be reported as empty also

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -435,6 +465,9 @@ public static boolean retainOrUpdateRecord(
       case MISSING:
         returnValue = container.isMissing() && !container.isEmpty();
         break;
+      case EMPTY_MISSING:
```
Pls explain the reason why we need to add this explicit `EMPTY_MISSING` switch case here because anyway we are deleting `EMPTY_MISSING` later at line [here]( https://github.com/apache/ozone/pull/7149/files#diff-5ecf43d8695ccad5a7f9d506546af1044b6034a2616787f8f881a276dc6f08fbR309)

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -273,34 +278,44 @@ private long processExistingDBRecords(long currentTime,
             LOG.info("DELETED existing unhealthy container record...for Container: {}",
                 currentContainer.getContainerID());
             rec.delete();
+            continue;
```
This can be removed because we want to proceed ahead for other unhealthy states. IMO once we remove EMPTY_MISSING switch case from there, we don't need this.

devmadhuu reviewed (commented):
Thanks @ArafatKhan2198 for this patch. Have few comments in line in code along with some general 2 comments below:

1. Pls correct the PR description "`when SCM marks a container as DELETED or identifies it as EMPTY_MISSING`", SCM never identifies unhealthy states.
2. Below check is invalid as we are removing EMPTY_MISSING containers from  `org.apache.hadoop.ozone.recon.fsck.TestContainerHealthTask#testRun`
```
UnHealthyContainers table:
LambdaTestUtils.await(6000, 1000, () -> {
      UnhealthyContainers emptyMissingContainer = unHealthyContainersTableHandle.fetchByContainerId(7L).get(0);
      return ("EMPTY_MISSING".equals(emptyMissingContainer.getContainerState()));
    });
```

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -435,6 +465,9 @@ public static boolean retainOrUpdateRecord(
       case MISSING:
         returnValue = container.isMissing() && !container.isEmpty();
         break;
+      case EMPTY_MISSING:
```
Thanks for pointing this out @devmadhuu 

This change of Negative Sized was added by this :- https://github.com/apache/ozone/pull/6178
The Negative sized containers do get added to the table, but they do have a bug in their implementation :- 

### Iteration 1 of `ContainerHealthTask`:

- **What happens**: The `checkAndProcessContainers` method fetches all containers, including the ones with negative sizes, since `processedContainers` is empty initially.
- **Result**: The negative-sized containers are processed and added to the `UNHEALTHY_CONTAINERS` table by `handleNegativeSizedContainers`.

### Iteration 2:

- **What happens**: The `processExistingDBRecords` method processes the previously added negative-sized containers. When it checks the records in the table, `retainOrUpdateRecord` returns `false` for negative-sized containers because the logic for keeping negative-sized records isn't present.
- **Result**: The negative-sized containers are deleted from the `UNHEALTHY_CONTAINERS` table, and the container is added to `processedContainers`. As a result, `checkAndProcessContainers` skips this container because it’s in the `processedContainers` set.

### Iteration 3:

- **What happens**: In this iteration, since there are no records for negative-sized containers in the `UNHEALTHY_CONTAINERS` table (due to their deletion in the previous iteration), the `processExistingDBRecords` method does not delete anything.
- **Result**: The negative-sized containers are not present in the table, so `checkAndProcessContainers` processes them again and adds them back to the `UNHEALTHY_CONTAINERS` table.

### Iteration 4:

- **What happens**: The process from Iteration 2 is repeated, where the `processExistingDBRecords` method deletes the negative-sized containers again because `retainOrUpdateRecord` still returns `false`.
- **Result**: The container gets deleted again, and the loop continues.

### Conclusion:

This behaviour creates a **cycle** where the negative-sized containers are added in one iteration and deleted in the next. This is due to the absence of logic in `generateUnhealthyRecords` to retain negative-sized containers. Adding proper handling for negative-sized containers in `generateUnhealthyRecords` should prevent this cycle.

devmadhuu replied:
Thanks @ArafatKhan2198  for your explanation above. My point is do we need to insert `NEGATIVE_SIZE` and `EMPTY_MISSING` containers in `UnhealthyContainers` table because they are anyway getting deleted in next iteration. There can be two ways to handle this:

1. Either DO NOT insert both these types containers in `UnhealthyContainers` table because they are getting deleted anyways in next iteration and we have logs to know the information on current count of both these types of containers with each task run. You can check the correctness of information in logs. This will make things simpler because no need to take care of constraint violation exception in case user moves to next higher version. And any existing MISSING type old containers which have actually zero keys needs to be cleaned up, so that Recon will not report such empty missing containers.
2. Or, Insert both these types of containers in `UnhealthyContainers` table which makes things little complex in terms of management of this data which is not being shown over Recon UI. Because such containers will continue to be lying in `UnhealthyContainers` table unless until we introduce some mechanism to clean up such containers periodically and then when and how to do cleanup of such containers makes things little overly complex.

By having solution 1, we also need to make sure about not to miss the corner edge case of updating container states of such containers after verification from SCM.

ArafatKhan2198 replied:
I agree @devmadhuu with the first solution! Based on my experience with customer issues, backward compatibility can be a significant challenge. We are anyways printing out the logs for it so no point in storing them.

sumitagrawl replied:
@ArafatKhan2198 @devmadhuu 
There is one more changes done, rename of one filed ALL_REPLICAS_BAD which also can cause compatibility issue.
IMO, current we can re-create constraint dropping and re-creating for this.

But we need support Schema upgrade feature for compatibility scenario for this DB.

ArafatKhan2198 commented:
@sumitagrawl @devmadhuu 

This PR updates how Recon handles MISSING, EMPTY_MISSING, and NEGATIVE_SIZE containers. The main change is that EMPTY_MISSING and NEGATIVE_SIZE containers will now only be logged in the statistics, but they will not be added to the unhealthyContainersTable.

The scenarios cover two situations:

- After Upgrade Scenarios: Describes how existing MISSING, EMPTY_MISSING, and NEGATIVE_SIZE containers will be handled after the upgrade, including when records are removed or logged.
- 
- Fresh Install Scenarios: Explains how these containers are managed on a new installation, focusing on proper logging without unnecessary database entries.

The goal is to make container health tracking more efficient by storing only necessary records and logging important details for containers that need attention.

---

### After Upgrade Scenarios:

1. **Existing DB records of containers `MISSING`:**
    - These containers remain in the `unhealthyContainersTable` as `MISSING` unless they are deleted in SCM. Their statistics will continue to be logged.
2. **Existing DB records of containers `MISSING` deleted in SCM:**
    - If SCM reports these containers as deleted, their state in `ReconContainerManager` will be changed to `DELETED`, and their records will be **removed** from the `unhealthyContainersTable`. No further logging of statistics will happen for these containers.
3. **Existing DB records of containers `MISSING` which are actually empty:**
    - These records will be logged under `EMPTY_MISSING` in the unhealthy container statistics and their **`MISSING`** record will be deleted from the `unhealthyContainersTable`. If SCM does not delete them, they continue to be logged as `EMPTY_MISSING` in subsequent iterations.
4. **Existing DB records of containers `MISSING` which are actually empty and also deleted in SCM:**
    - If SCM reports them as deleted, their state in Recon will be updated to `DELETED`, and the records will be **removed** from the `unhealthyContainersTable`. No further logging will occur for these containers.
5. **New `MISSING` containers added:**
    - These containers will be added to the `unhealthyContainersTable` as `MISSING` and their statistics logged.
6. **New `MISSING` containers added which are also deleted in SCM:**
    - If these containers are reported as deleted in SCM, their state will transition to `DELETED` in Recon, and their records will be **removed** from the `unhealthyContainersTable`. No further statistics will be logged.
7. **New**  `EMPTY_MISSING` **containers added:**
    - These containers will be logged under `EMPTY_MISSING` in the unhealthy container statistics, but they will **not** be added to the `unhealthyContainersTable`.
8. **New** `EMPTY_MISSING` **containers added which are also deleted in SCM:**
    - If SCM reports them as deleted, they will not be added to the `unhealthyContainersTable`, and their state will be transitioned to `DELETED` in Recon. No further logging will happen.
9. **New `NEGATIVE_SIZE` containers added:**
    - These containers will be logged under `NEGATIVE_SIZE` but **not** added to the `unhealthyContainersTable`. They will only appear in the unhealthy container statistics and no further records will be stored in the table.

---

### Fresh Install Scenarios:

1. **`MISSING` containers which are deleted in SCM:**
    - These containers will have their state updated to `DELETED` in Recon, and no records will be added to the `unhealthyContainersTable`.
2. **`MISSING` containers which are not deleted in SCM:**
    - These containers will be added to the `unhealthyContainersTable` as `MISSING` and will have their statistics logged.
3. `EMPTY_MISSING` **containers which are deleted in SCM:**
    - These containers will not be added to the `unhealthyContainersTable`, and their state will be updated to `DELETED` in Recon. No further logging will happen.
4. `EMPTY_MISSING` **containers which are not deleted in SCM:**
    - These containers will be logged under `EMPTY_MISSING` in the unhealthy container statistics but **not** added to the `unhealthyContainersTable`.
5. **`MISSING` containers which get back some replicas reported:**
    - These containers will transition from `MISSING` to a healthy state. The record will be updated or removed from the `unhealthyContainersTable` accordingly.
6. **`NEGATIVE_SIZE` containers added:**
    - These containers will be logged under `NEGATIVE_SIZE` but **not** added to the `unhealthyContainersTable`. Only the statistics for these containers will be tracked in the unhealthy container stats, and no records will be stored.

ArafatKhan2198 commented:
I needed to ensure that the code for altering the table works correctly.

- For testing purposes, I made some changes to the code where I drop the existing constraint and add a new one that allows all unhealthy container states except MISSING. Please note, the removal of the MISSING container state is temporary and only for testing purposes; I will add it back later.
- The reason for this test is that, during the initial load, Recon creates a schema that accepts all container states.
- Upon restarting Recon, it checks if the table exists. Since the table will already exist after the initial load, the system will drop the current constraint and add a new one where the MISSING container state is excluded for testing purposes.

- Below are the code changes I made to the method:

```
  private void addUpdatedConstraint() {
    // Get all enum values as a list of strings
    String[] enumStates = Arrays.stream(UnHealthyContainerStates.values())
        .map(Enum::name)
        .filter(state -> !state.equals("MISSING"))  // Exclude MISSING for testing
        .toArray(String[]::new);

    LOG.info("Adding updated constraint with states: {}", Arrays.toString(enumStates));

    // Alter the table to add the updated constraint
    dslContext.alterTable(UNHEALTHY_CONTAINERS_TABLE_NAME)
        .add(DSL.constraint(UNHEALTHY_CONTAINERS_TABLE_NAME + "ck1")
            .check(field(name("container_state"))
                .in(enumStates)))
        .execute();
  }
  
```
To test this, I simulated MISSING containers in Recon by creating a container and then stopping all the datanodes, which triggers the MISSING container scenario.
When Recon attempted to add these MISSING container records to the unhealthyContainerTable, it failed and threw a constraint violation error, confirming that our code works correctly.

```
2024-09-10 18:09:23 2024-09-10 12:39:23,070 [Recon-EventQueue-DeadNodeForReconDeadNodeHandler] ERROR scm.ReconDeadNodeHandler: Error trying to verify Node operational state from SCM.
2024-09-10 18:09:23 org.jooq.exception.DataAccessException: SQL [insert into "UNHEALTHY_CONTAINERS" ("container_id", "container_state", "in_state_since", "expected_replica_count", "actual_replica_count", "replica_delta", "reason") values (cast(? as bigint), cast(? as varchar(32672)), cast(? as bigint), cast(? as int), cast(? as int), cast(? as int), cast(? as varchar(32672)))]; The check constraint 'UNHEALTHY_CONTAINERSck1' was violated while performing an INSERT or UPDATE on table '"RECON"."UNHEALTHY_CONTAINERS"'.
```

This error shows that the `MISSING` container state is not allowed, as expected, confirming that the new constraint works correctly. I will later reintroduce the MISSING container state into the list once testing is complete

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -273,34 +278,44 @@ private long processExistingDBRecords(long currentTime,
+          //  Unhealthy Containers such as MISSING, EMPTY_MISSING, UNDER_REPLICATED,
+          // OVER_REPLICATED, MIS_REPLICATED can have their unhealthy states changed or retained.
+          if (!ContainerHealthRecords.retainOrUpdateRecord(currentContainer, rec)) {
             LOG.info("DELETED existing unhealthy container record...for Container: {}",
```
Done!

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -273,34 +279,41 @@ private long processExistingDBRecords(long currentTime,
+
+          existingRecords.add(rec.getContainerState());
+          // If the record was changed, update it
+          if (rec.changed()) {
```
there is change in behavior as compared old if-else. earlier for one case, rec.changed() and this case is never called for else. But now its called both case, please verify the impact.
cc: @devmadhuu

ArafatKhan2198 replied:
This is what I understand :- 

`rec.delete()`: This method directly deletes the record from the database by issuing a SQL DELETE statement. It doesn’t modify the record's fields or mark them as "changed." Instead, it removes the corresponding row from the database. Once deleted, the record is no longer considered part of the database, and calling rec.changed() afterward would not make sense because there is no row left to update.

`rec.changed()`: This method is used to detect whether any fields in the record have been modified since it was fetched or created. Jooq internally tracks changes to the record fields (e.g., setReplicaDelta(), setContainerState()), and if any field has been modified, rec.changed() will return true. This helps Jooq decide whether an UPDATE query needs to be sent to the database when rec.update() is called.

In jOOQ, when you call rec.delete(), it removes the record from the database, and this operation is executed independently of the rec.changed() flag. The changed() method in jOOQ is generally used to determine if any fields in the record have been modified, and it's primarily for UPDATE operations. When a record is deleted using delete(), jOOQ executes a DELETE statement, and this does not affect the record's changed() state.

But I will verify this once by testing.

cc: @devmadhuu

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java:
```diff
@@ -394,14 +422,9 @@ private void handleNegativeSizedContainers(
             "container endpoint for more details.",
         container.getContainerID());
-    UnhealthyContainers record =
-        ContainerHealthRecords.recordForState(containerHealthStatus,
```
need remove handleNegativeSizedContainers() as not used

ArafatKhan2198 replied:
Actually I checked again it is being used by processContainer() method, that runs after we process the existing DB Records.

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java:
```diff
@@ -160,6 +160,10 @@ public boolean isEmpty() {
     return numKeys == 0;
   }
 
+  public boolean isEmptyMissing() {
```
Need remove this method, not used

ArafatKhan2198 replied:
Done

devmadhuu reviewed (commented):
@ArafatKhan2198 thanks for the patch, Pls see some comments.

sumitagrawl reviewed (approved):
LGTM

devmadhuu commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconTasks.java:
```diff
@@ -272,7 +285,17 @@ public void testEmptyMissingContainerDownNode() throws Exception {
+      return allEmptyMissingContainers.size() == 0 &&
+          unhealthyContainerStateStatsMap.get(
+                  ContainerSchemaDefinition.UnHealthyContainerStates.EMPTY_MISSING)
+              .getOrDefault(CONTAINER_COUNT, 0L) == 0;
```
Shouldn't we check the `CONTAINER_COUNT` for `MISSING` state after we added keys in container ? We did already for EMPTY_MISSING above.. Let's do in next PR.

devmadhuu reviewed (approved):
Thanks @ArafatKhan2198 for the patch. LGTM +1.
